### DISCUSSION TITLE
How to make player stop when colliding with object?

### DISCUSSION THREAD

**Discussion by mattoni** [jcornaz/heron#222] [Q&A] [answered]
I'm trying to create a simple 2d rpg style game. I've turned off gravity and simply am trying to get my player character to collide and stop when hitting an object. The current behavior is they glide through, but will slowly "slide" off the object if they are collided.

The player is set up like so:

```
 commands
        .spawn_bundle(PlayerBundle {
            sprite: SpriteSheetBundle {
                texture_atlas: texture_atlas_handle,
                transform: Transform::from_scale(Vec3::splat(1.0)),
                sprite: TextureAtlasSprite::new(avatar.id * 3),
                ..Default::default()
            },
            movement: Movement {
                is_moving: false,
                direction: Direction::Down,
            },
        })
        .insert(RigidBody::Dynamic)
        .insert(CollisionShape::Cuboid {
            half_extends: Vec2::new(32., 32.).extend(0.0) / 2.0,
            border_radius: None,
        })
```
and the static object like this:

```
    commands
        .spawn_bundle(SpriteBundle {
            texture: asset_server.load("image.png"),
            ..Default::default()
        })  
        .insert(RigidBody::Static)
        .insert(CollisionShape::Cuboid {
            // let the size be consistent with our sprite
            half_extends: Vec2::new(32., 32.).extend(0.0) / 2.0,
            border_radius: None,
        });
```

To move my character, I use something like this: 

```
pub fn player_movement_system(mut q: Query<(&Movement, &mut Transform), With<Player>>,) {
    let (movement, mut transform) = q.single_mut();

    match movement {
        Movement {is_moving: true, direction: Direction::Up} => {
            transform.translation.y += 1.;
        },
        Movement {is_moving: true, direction: Direction::Down} => {
            transform.translation.y -= 1.;
        },
        Movement {is_moving: true, direction: Direction::Left} => {
            transform.translation.x -= 1.;
        },
        Movement {is_moving: true, direction: Direction::Right} => {
            transform.translation.x += 1.;
        },
        _ => ()
    }
}
```
Any suggestions on how to properly configure the physics for a basic interaction like this? Ideally this should work without me needing to do manual collision detection :/ 

Thanks!

**Comment by jcornaz** [comment, marked as answer]
Updating the transform as you do, is *teleporting* the object each frame, ignoring all physics rules.

If you want a more realistic behavior (like not passing through other objects :-p) then it's more appropriate to use the velocity component.

See: https://docs.rs/heron/latest/heron/#move-rigid-bodies-programmatically

**Reply by mattoni** [reply]
Thanks! I switched to using the velocity, which works, however the player kept going even without keydown unless it runs into the item. So to counter, I set velocity to 0 when the movement keys aren't being pressed, which gives me much more similar interaction to a classic pixel game. However, one final strange issue occurs - I can sort of 'slide' next to the object and it seems like it has some friction, even tho I set friction on the static body to 0. My character basically gets somewhat stuck and holding down the key eventually frees them. Is there some simple way to make the object just act like a static barrier with no other interaction?

Thanks!

**Reply by mattoni** [reply]
Turns out, removing friction completely from the player resolved it.
